### DISCUSSION TITLE
Mac OS Integration Questions - App Bundle

### DISCUSSION THREAD

**Discussion by yveszoundi** [fltk-rs/fltk-rs#1212] [Q&A] [answered]
I'm working on a [document sanitization tool](https://github.com/rimerosolutions/dangerzone-rust) that converts potentially suspicious files to PDFs. I'm trying to provide a DMG installer and an app bundle, instead of double-clicking on an executable that pops-up a _Terminal_ window.

**Accepting file arguments for an application bundle**
My current understanding is that under MacOS dispatches "[Apple Events](https://stackoverflow.com/questions/56532571/passing-arguments-to-macos-app-when-bundle-executable-is-a-bash-script)" that need to be handled by the application.
Essentially, I want to allow users to open files from the _Finder_ file manager ("Open With").
- I'm hoping that FLTK already provides such integration for Mac OS, is it the case?
- Otherwise it seems that I need some kind of callback registration possibly to the apple event manager (maybe something similar to the [fruitbasket](https://docs.rs/crate/fruitbasket/latest/source/src/osx.rs) library).

**Launching other programs outside the Mac OS sandbox**
My applications relies on invoking external programs such as Docker or equivalent solutions. It appears that the only way for it to work in my application is to wrap the UI executable within a shell script (hard-coding executable paths doesn't work). 
Is there a better way from your experience?

```bash
#!/bin/bash
abspath () 
{ 
case "${1}" in 
    [./]*)
    local ABSPATH="$(cd ${1%/*}; pwd)/"
    echo "${ABSPATH/\/\///}"
    ;;
    *)
    echo "${PWD}/"
    ;;
esac
}

CURRENTPATH=`abspath ${0}`
PATH=$PATH:$CURRENTPATH

# passing arguments doesn't work this way inside app bundles per first question
ARGS="$*"

$SHELL -l -c "dangerzone-gui ${ARGS}"
```

**Comment by MoAlyousef** [comment, marked as answer]
Hi

1- FLTK offers a callback which is triggered when a file is dropped on the app bundle on macos:
https://www.fltk.org/doc-1.4/group__group__macosx.html#ga0702a54934d10f5b72157137cf291296

This is currently not exposed in fltk-rs, but it's easy to add.

There is also (the unsafe) [`app::add_system_handler()`](https://docs.rs/fltk/latest/fltk/app/fn.add_system_handler.html) which takes a callback acting on a system NSEvent on macos. But if I understand correctly, these seem to be different from "Apple Events".

2- AFAIR launching apps outside of your app's sandbox is not possible, wrapping your app in a shell script probably defeats the idea of creating an app bundle in the first place. That said, I'm not experienced with mac programming, so it might be possible to configure such a feature in the info.plist or the entitlements.plist.

Maybe some other developers can offer insights here, you can also ask on the [apple developers forums](https://developer.apple.com/forums/) or the [fltk-general](https://groups.google.com/g/fltkgeneral) mailing list.

**Comment by yveszoundi** [comment]
Thanks for the prompt response @MoAlyousef , 

1. After more research, I think that support for "apple open events" is already in `fltk-rs` itself, but not exposed (`fltk/src/app/event.rs`). It would be great to make it publicly available in future releases.
https://github.com/fltk-rs/fltk-rs/blob/e7bbe682697e8421decadb3371b9fad044667f1c/fltk/src/app/event.rs#L549 

2. I'm more familiar with Linux than Mac OS myself, but I'm still exploring better options (made it "work" first at least).
- Most Apple users likely expect a `DMG image` containing an app bundle for GUI programs: I suspect that some people even find it "totally unacceptable" not to have an app bundle for a graphical application on the Mac.
- While potentially unusual, I think that it's not necessarily too uncommon to leverage external "known programs" within an app. Again, I'm still searching in forums, mailing lists, etc.

**Reply by MoAlyousef** [reply]
1- I've added this function after seeing this current discussion. Note that it's now:
```rust
pub fn raw_open_callback(cb: Option<fn(*const raw::c_char)>)
```
because I can't find a way to seamlessly convert a function pointer with a const char * argument to a function pointer/object with a &str or Path argument.
Also it might be easier if one needs to call c/objective-c api.
So it can currently be called:
```rust
use fltk::{app, dialog};
app::raw_open_callback(Some(|s| {
   let file_path = unsafe { std::ffi::CStr::from_ptr(s).to_string_lossy().to_string() };
   dialog::message_default(&format!("You dropped {}", file_path));
}));
```

2- I agree. Currently this limitation comes from Rust's tooling. For example, with CMake, you can pass MACOSX_BUNDLE to the add_executable() call and it will create an app bundle, also you can create a DMG using CMake's CPack module.

With cargo, you need to install cargo-bundle to create an app bundle, but it doesn't support creating DMG images.
Also both cargo and rustc don't support universal macos fat binaries (targetting both x86_64 and arm64 (new m1 chip)).

**Comment by yveszoundi** [comment]
Thanks again,

1. It  seems that I would not be able to "process" the file while referencing other variables. Is it technically possible to have a closure callback?

I'd like to reference my custom widget `filelist_widget` per excerpt below or somehow dispatch custom events (along with file data). 

```rust
if cfg!(target_os = "macos") {        
        app::raw_open_callback({
            let mut filelist_widget_ref = filelist_widget.clone();
            
            Some(|s| {                    
                let input_path: String = {
                    let ret = unsafe { std::ffi::CStr::from_ptr(s).to_string_lossy().to_string() };
                    ret.to_owned()
                };
                // I get the expected error `| |_____________^ expected fn pointer, found closure`.
                filelist_widget_ref.add_file(PathBuf::from(input_path));
            })
        });
}
```

2. I do have a flexible approach via [docker-create-dmg](https://github.com/sporsh/docker-create-dmg), but thanks reminding me of `cargo-bundle`.

**Reply by MoAlyousef** [reply]
FLTK's fl_open_callback function doesn't take an extra void * argument, so we can't use that to pass a capturing function object unfortunately.

So there are 2 approaches around this:
- Using channels:
```rust
use fltk::{app, frame::Frame, prelude::*, window::Window};

fn main() {
    let app = app::App::default();
    let (_, r) = app::channel::<String>();
     app::raw_open_callback(Some(|s| {
        let input_path: String = {
            let ret = unsafe { std::ffi::CStr::from_ptr(s).to_string_lossy().to_string() };
            ret.to_owned()
        };
        let s = app::Sender::<String>::get();
        s.send(input_path);
     }));
    let mut wind = Window::default().with_size(400, 300);
    let mut frame = Frame::default().with_size(200, 100).center_of_parent();
    wind.end();
    wind.show();

    while app.wait() {
        if let Some(msg) = r.recv() {
            frame.set_label(&msg);
        }
    }
}
```

- Using statics (via a safe wrapper like lazy_static for example):
```rust
use fltk::{app, frame::Frame, prelude::*, window::Window};

lazy_static::lazy_static! {
    static ref FRAME: Frame = Frame::default();
}

fn main() {
    let app = app::App::default();
     app::raw_open_callback(Some(|s| {
        let input_path: String = {
            let ret = unsafe { std::ffi::CStr::from_ptr(s).to_string_lossy().to_string() };
            ret.to_owned()
        };
        FRAME.clone().set_label(&input_path);
     }));
    let mut wind = Window::default().with_size(400, 300);
    let mut frame = FRAME.clone().with_size(200, 100).center_of_parent();
    wind.end();
    wind.show();

    app.run().unwrap();
}
```
The [widget_id example](https://github.com/fltk-rs/fltk-rs/blob/master/fltk/examples/widget_id.rs) stores widgets into a lazy_static WIDGET_MAP and accesses them via an id (like in html). You can take a look at that as well.

**Reply by yveszoundi** [reply]
It works beautifully, I went ahead with the channel approach.
